VM snapshots (#428)
openshift/openshift-origin-design: designs/virtualization/4.x/snapshots/snapshots.md

---
parent: Virtualization
version: 4.x
---

# VM snapshots

This item will cover updates to vm-snapshots
- Creating a VM snapshot
- Snapshots on the VM details page
- Virtualization snapshot list
- Restoring a VM from snapshot

## Creating a snapshot

![Virtualization - VMs list](img/0-0.jpg)
From the virtual machines menu item, the user selects “Create Snapshot” from a VM item’s kebab dropdown.
The same action can be taken from a VM details Action dropdown.

> ![Virtualization - turn off VM notification](img/0-1.jpg)

Reviewer: Can we confirm that the icon is meant to be at the top like this and not beside the content of the body? In the conventions it shows the latter.

Author: @itsptk can you confirm that the icon is meant to be on the top next to the title?
I believe I saw this pattern on the operators' domain
Thanks

Reviewer: ah yes I see it for the operator here when it's being uninstalled
I do question though if this is the way it will be implemented when the modals from PF4 get pulled in. They don't current have support for adding icons to the heading like this (But we could ask for it). @itsptk have you talked to the PF side about this at all?
<img width="641" alt="Screen Shot 2020-07-08 at 4 43 43 PM" src="https://user-images.githubusercontent.com/287880/86963030-3f26e900-c13a-11ea-88c0-18f088ef01e7.png">

Reviewer: follow up - looks like this is on the books for PF. So we should be ok to continue with these designs.
https://github.com/patternfly/patternfly-design/issues/863
Seems like a nice addition to the alert convention :)

Reviewer 2: Sorry for delay but yeah, as that PF issue mentions it was the recommended method in the design guidelines for destructive confirmation modals: https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/modal#confirm-a-destructive-action

In case the VM is running, the user will be asked to turn it off to make a snapshot

> ![Virtualization - Create VM snapshot](img/1-0.jpg)

Reviewer: Do you think we need the description for the VM here?

Author: I'm not sure. I thought that it is not required since the user had selected a VM just want to make sure that while on the Create screen, it is the VM he had chosen. But I'm not against adding a description. WDYT?

Reviewer: ya I'm actually saying do we need to even show the description at all. I'm wondering how useful it is for a snapshot.

Author: I think it comes down to the question if we think the user will ever need to double-check which item she is doing that (full page form) action on, not just snapshots.

I agree that it is very unlikely to actually take an action on an item and then forget what that item was.
BUT it is a full-page and we need to mention what is the subject of the page.

So this leads me to the question, is the name of the subject (workload) is enough to make it recognizable?

I think that we do need to add the name just below the title, for smaller screens.
I don't see how the right side column does any damage. It is a nice touch.
Having said that, it is a dev effort and we do need to make it a design pattern, so it is hard for me to evaluate the benefit vs. the effort.

Author: This discussion will be pushed to a future design. For now, the current implementation was using a modal so no "Creation" page is included.

The user enters the Create snapshot screen.
This screen includes:
- Snapshot name field

> - Save memory content (future implementation)

Reviewer: can we separate out these "Future implementation" items? It would be good to have it assigned to 4.x so we don't forget about it.

Reviewer: ah I just realized these are all under 4.x so maybe we don't need to say these are future implementations? OR if we don't think it will get in whenever its implemented we should provide designs that cover what will go in first. WDYT?

Author: I think I better drop it. 
As for the Included disks @aglitke mentioned it is not implemented yet. but the disks are included in the snapshot.
I think it is ok to keep that dropdown and drop the "Future version" from it, but I'm waiting to hear which version will this dropdown go into.

Author: @matthewcarleton So the snapshots should be for 4.6 but it may be ready for 4.7.
So I am placing it ina 4.7 for now.
I also removed the "Included disks" drop-down and the "Include memory" checkbox and place it as a new MD in 4.x.
I'm using this PR for both MDs. Let me know if that's ok.

Reviewer: perfect! so @jelkosz was saying today that we will see *something* for 4.6. Do we know what that is? Should we just include that for the 4.6 and then just put everything else in 4.x for now?

- Disks included multi-select drop-down (future implementation)

> - Dropdown to selecting Snapshot Class

Reviewer: should we explain why this is auto selected?

Author: Explanation added

- A summary of the VM details on the right-side column
- A summary of the size of the snapshot (above the primary action)

> ### Selecting disks (future implementation)

Reviewer: move to new 4.x file?

> ![Virtualization - "un-snapshotable" disks](img/1-0-b.jpg)

Reviewer: is it possible to provide a reason why they can't be in the snapshot?

Author: I believe the new info box covers this.

> In case there are some disks who cannot be "snapshotted" an in-line notification will notify that and specify those disks.

Reviewer: ```suggestion
In case there are some disks which cannot be "snapshotted" an in-line alert will warn that and specify those disks.
```

Author: Fixing this in the 4.x MD

> ![Virtualization - disks dropdown](img/1-1.jpg)

Reviewer: it doesn't look like we have this type of select in PatternFly - have you raised this on the PF side? I think we can cover this with an existing solution in PF WDYT?

Author: This is no longer relevant, right (can't see the images)

Selecting a disk. All disks are selected by default.
The "Un-snapshotable" disks will also be previewed within the disk multi-select drop-down, at a separate section at the bottom of the dropdown as a read only.

![Virtualization - Selecting a disk](img/1-2.jpg)
Selecting a specific disk will add that disk to a multi-select preview within the field.

### Creating screen

> ![Virtualization - Creating snapshot](img/2-0.jpg)

Reviewer: I think we can align this screen with what @itsptk did here
http://openshift.github.io/openshift-origin-design/designs/administrator/olm/operand-creation-at-operator-install/
@beanh66 correct me if I'm wrong :)

Reviewer 2: Ran and I had discussed perhaps aligning the two areas around that design. That said, I think it could probably have its hierarchy tweaked to lend itself to more use cases (not just installing operators) if needed.

Reviewer 2: Component for this purpose is being sorted out here: https://docs.google.com/document/d/1RrtGWMkoXLRnKuDzWApuOBQn9kCpnhQDfzrz-KNB-r4/edit?usp=sharing

After clicking the Create button, the user goes to a transition page, in “Creating mode”.
While the snapshot is being created the user can navigate external to this flow. A text below will mention that, with a link to going back.

## VM details > Snapshots

![Virtualization - Snapshot details page](img/2-1.jpg)
In case creation succeeded, the user will land on the snapshot details page.
The snapshot details page should contain all the data relevant to the VM.
Overview tab is not available.

![VM details - Snapshots list](img/3-0.jpg)
All the columns are sortable within the snapshot table.

![VM details - Snapshots list](img/2-2.jpg)
Snapshots will also be surfaced on the VM details > Overview > Inventory

## Virtualization > Snapshots list

![Virtualization - Snapshots list page](img/5-0.jpg)
This list will include Parent (VMs).

> ![Virtualization - Snapshots list - filters](img/5-01.jpg)

Reviewer: I'm sure I'm missing something here - This feels a bit confusing to me. Is the filter filtering the parent? I don't think this has been implemented yet on the search page right? Maybe we put this in 4.x too?

Author: Moving this to 4.x MD

Multi-select filter (like we have on the search page)

# Restoring VM from snapshot

![Virtualization - restore from snapshot](img/6-0.jpg)
The user selects the restore action.

> ![Virtualization - restore from snapshot](img/6-1.jpg)

Reviewer: same comment about icon here

Author: Moving the "take a snapshot of the VM you are about to restore" to 4.x

In case the VM is running, a warning message will mention that restoring for a running VM is not possible and that VM should be stopped first.

> ![Virtualization - restore from snapshot](img/6-2.jpg)

Reviewer: nit - can we provide this type of default naming format? I'm not sure about the space before the number at the end.

Author: Adding "-" between the spaces

The user enters a “Review and Restore” screen.

> ![Virtualization - restore from snapshot](img/6-3.jpg)

Reviewer: If we are moving this future stuff to 4.x we should provide designs for what will be implemented first.

Author: I've split the design into two different MDs (covered in this PR)
One for 4.7 and one for 4.x
